src/OrchardCore.Modules/OrchardCore.Deployment/Controllers/ImportController.cs

```diff
@@ -102,7 +102,54 @@ public async Task<IActionResult> Import(IFormFile importedPackage)
                 _notifier.Error(H["Please add a file to import."]);
             }
 
-            return RedirectToAction("Index");
+            return RedirectToAction(nameof(Index));
+        }
+
+        public async Task<IActionResult> Json()
+        {
+            if (!await _authorizationService.AuthorizeAsync(User, Permissions.Import))
+            {
+                return Forbid();
+            }
+
+            return View();
+        }
+
+        [HttpPost]
+        public async Task<IActionResult> Json(string json)
+        {
+            if (!await _authorizationService.AuthorizeAsync(User, Permissions.Import))
+            {
+                return Forbid();
+            }
+
+            if (!string.IsNullOrWhiteSpace(json))
```

Comment:
you could use that `IsJson()` extension I added here, and rather than a notifier add an error to the model state.

Then check if the model state is valid, and `View(json)` to return the bad value to the form.
It should probably be a class
`ImportJsonViewModel`

Then your binding in the view can use `asp-for` instead of the manual id and name you have.

Should end up looking something like this
```
            if (ModelState.IsValid)
            {
                ...run the recipe...

                return RedirectToAction(nameof(Json));
            }

            // If we got this far, something failed, redisplay form
            return View(model);
```

Reply:
ok